### PR TITLE
[Parallel Algorithms] Make R3 for P3179 Parallel Range Algorithms

### PR DESCRIPTION
**Pull request by rarutyun**

### CODE REVIEW

File: parallel_algorithms/P3179/P3179.bs [hunk 1]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+                           Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
+template<class ExecutionPolicy, random_access_range R1, random_access_range R2, class Pred = ranges::equal_to,
+         class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2> && (sized_range<R1> || sized_range<R2>)
+  bool ranges::starts_with(ExecutionPolicy&& exec, R1&& r1, R2&& r2, Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
+</xmp>
+
+<i>Returns</i>:
+    `ranges::mismatch(std::forward<ExecutionPolicy>(exec), std::move(first1), last1, std::move(first2), last2,
+                          pred, proj1, proj2).in2 == last2`
```
Comment (akukanov) [inline parallel_algorithms/P3179/P3179.bs:1276]: I would double-check that use of `||` for `sized_range` requirements is right for this algorithm. The concern is about `mismatch(...).in2==last2` for the return value. If the second range `[first2,last2)` is unbounded, will the returned iterator to the range ever be equal to `last2`, even if all elements of the first range were matched?

Or in other words, does not the semantics of `starts_with` assume that the second sequence is finite?

File: parallel_algorithms/P3179/P3179.bs [hunk 2]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+!!! TODO: review the constrains carefully !!!
+<ins>
+<xmp class="wordingStyle">
+template<class ExecutionPolicy, random_access_iterator I, sized_sentinel_for<I> S, random_access_iterator O,
+         class Proj = identity,
+         class T = projected_value_t<I, Proj>, indirect_unary_predicate<projected<I, Proj>> Pred>
+  requires indirectly_copyable<I, O> && indirectly_writable<O, const T&>
+  ranges::replace_copy_if_result<I, O>
+    ranges::replace_copy_if(ExecutionPolicy&& exec, I first, S last, O result,
+                            Pred pred, const T& new_value, Proj proj = {});
```
Comment (akukanov) [inline parallel_algorithms/P3179/P3179.bs:1651]: You change the default type for `T` from being defined by output (`T = iter_value_t<O>`) to being defined by input (`T = projected_value_t<I, Proj>`). Is it an oversight or a deliberate choice?

File: parallel_algorithms/P3179/P3179.bs [hunk 3]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+template<class ExecutionPolicy, random_access_range R, random_access_iterator O, class Proj = identity,
+         indirect_unary_predicate<projected<iterator_t<R>, Proj>> Pred>
+  requires indirectly_copyable<iterator_t<R>, O> && sized_range<R>
+  ranges::remove_copy_if_result<borrowed_iterator_t<R>, O>
+    ranges::remove_copy_if(ExecutionPolicy&& exec, R&& r, O result, Pred pred, Proj proj = {});
+</xmp>
+</ins>
+
+## Modify `unique` in <a href="https://eel.is/c++draft/alg.unique">[**alg.unique**]</a> ## {#modify_unique}
+!!!!! TODO: REVIEW CONSTRAINTS FOR UNIQUE_COPY !!!!!!!
```
Comment (akukanov) [inline parallel_algorithms/P3179/P3179.bs:1836]: I took a look, and I think `unique_copy` is good.

File: parallel_algorithms/P3179/P3179.bs [hunk 4]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+template<class ExecutionPolicy, random_access_range R1, random_access_range R2, class Pred = ranges::equal_to,
+         class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2> && (sized_range<R1> || sized_range<R2>)
+  bool ranges::equal(ExecutionPolicy&& exec, R1&& r1, R2&& r2,
+                     Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
+</xmp>
+</ins>
+
+## Modify `search` in <a href="https://eel.is/c++draft/alg.search">[**alg.search**]</a> ## {#modify_alg_search}
+
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1220]: That's absolutely true :)

File: parallel_algorithms/P3179/P3179.bs (hunk 3 above)
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1941]: Thanks. TODO removed (locally, so far)

File: parallel_algorithms/P3179/P3179.bs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1274]: Done.

File: parallel_algorithms/P3179/P3179.bs (hunk 1 above)
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1380]: I think we agreed on "both sized"

File: parallel_algorithms/P3179/P3179.bs [hunk 5]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+                               Proj1 proj1 = {}, Proj2 proj2 = {});
+</xmp>
+
+<ins>
+<xmp class="wordingStyle">
+template<class ExecutionPolicy, random_access_iterator I1, sized_sentinel_for<I1> S1, random_access_iterator I2, sized_sentinel_for<I2> S2,
+         class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<I1, I2, Pred, Proj1, Proj2>
+  bool ranges::equal(ExecutionPolicy&& exec, I1 first1, S1 last1, I2 first2, S2 last2,
+                     Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1263]: Done.

File: parallel_algorithms/P3179/P3179.bs [hunk 6]
```diff
@@ -713,14 +727,2242 @@ Here are questions we want to answer (potentially not a complete list):
...
+         class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<I1, I2, Pred, Proj1, Proj2> && (sized_sentinel_for<S1, I1> || sized_sentinel_for<S2, I2>)
+  ranges::mismatch_result<I1, I2>
+    ranges::mismatch(ExecutionPolicy&& exec, I1 first1, S1 last1, I2 first2,S2 last2,
+                     Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
+template<class ExecutionPolicy, random_access_range R1, random_access_range R2,
+         class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2> && sized_range<R1> && sized_range<R2>
+  ranges::mismatch_result<borrowed_iterator_t<R1>, borrowed_iterator_t<R2>>
+    ranges::mismatch(ExecutionPolicy&& exec, R1&& r1, R2&& r2, Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1237]: Done.

File: parallel_algorithms/P3179/P3179.bs (hunk 2 above)
Comment (rarutyun) [reply parallel_algorithms/P3179/P3179.bs:1755]: An oversight, thanks.

**Review by akukanov — APPROVED** [review]

**Pull request merged**